### PR TITLE
python3Packages.jaxlib: fix darwin build

### PR DESCRIPTION
**Pull request by uri-canva**
###### Description of changes

Fix the darwin build for jaxlib.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by uri-canva** [comment]
@ofborg build python39Packages.jaxlib

**Comment by uri-canva** [comment]
@ofborg build python3Packages.jaxlib

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 1]
```diff
@@ -251,13 +265,19 @@ let
 
     inherit meta;
   };
+  platformTag =
+    if stdenv.targetPlatform.isLinux then
+      "manylinux2012_${stdenv.targetPlatform.linuxArch}"
```
Comment (uri-canva) [inline pkgs/development/python-modules/jaxlib/default.nix:270]:
```suggestion
      "manylinux2010_${stdenv.targetPlatform.linuxArch}"
```

**Comment by uri-canva** [comment]
Build timed out: https://logs.nix.ci/?attempt_id=96c8179f-ec2f-44ea-b029-ef19bdc1255b&key=nixos%2Fnixpkgs.184395

**Comment by samuela** [comment]
I don't have x86_64-darwin to test, but I can try on arch64-darwin

**Comment by uri-canva** [comment]
On ofborg it failed to build `aarch64-darwin`, and not due to timeout, some table gen issue.

**Comment by uri-canva** [comment]
Ah no I remembered it wrong, it was because it can't find a `darwin_arm64` toolchain: https://logs.nix.ci/?attempt_id=4a181b56-e474-4456-8ac7-47daf859ad27&key=nixos%2Fnixpkgs.184395, probably something to do with the different ways M1 identifies itself, I think I've hit it before. Sorry I didn't want to look more into `aarch64-darwin` because mine is a 13", so the builds take forever.

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 2]
```diff
@@ -199,26 +203,24 @@ let
       "zlib"
     ];
 
-    # Make sure Bazel knows about our configuration flags during fetching so that the
-    # relevant dependencies can be downloaded.
-    bazelFetchFlags = bazel-build.bazelBuildFlags;
-
-    bazelBuildFlags = [
```
Comment (samuela) [inline pkgs/development/python-modules/jaxlib/default.nix:206]: Why was this removed and merged into a single `bazelFlags`? It looks like we're losing `bazelFetchFlags` entirely?

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 3]
```diff
@@ -236,13 +238,25 @@ let
...
+        substituteInPlace ../output/external/rules_cc/cc/private/toolchain/unix_cc_configure.bzl \
+          --replace "/usr/bin/libtool" "${cctools}/bin/libtool"
+        substituteInPlace ../output/external/rules_cc/cc/private/toolchain/osx_cc_wrapper.sh.tpl \
+          --replace "/usr/bin/install_name_tool" "${cctools}/bin/install_name_tool"
+      '' + (if stdenv.cc.isGNU then ''
+        sed -i 's@-lprotobuf@-l:libprotobuf.a@' ../output/external/org_tensorflow/third_party/systemlibs/protobuf.BUILD
+        sed -i 's@-lprotoc@-l:libprotoc.a@' ../output/external/org_tensorflow/third_party/systemlibs/protobuf.BUILD
+      '' else if stdenv.cc.isClang then ''
+        sed -i 's@-lprotobuf@${pkgs.protobuf}/lib/libprotobuf.a@' ../output/external/org_tensorflow/third_party/systemlibs/protobuf.BUILD
+        sed -i 's@-lprotoc@${pkgs.protobuf}/lib/libprotoc.a@' ../output/external/org_tensorflow/third_party/systemlibs/protobuf.BUILD
```
Comment (samuela) [inline pkgs/development/python-modules/jaxlib/default.nix:258]: Why do we link against `pkgs.protobuf` when using clang, but the vendored version when building with gcc?

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 4]
```diff
@@ -251,13 +265,19 @@ let
 
     inherit meta;
   };
+  platformTag =
+    if stdenv.targetPlatform.isLinux then
+      "manylinux2010_${stdenv.targetPlatform.linuxArch}"
+    else if stdenv.targetPlatform.isDarwin then
+      "macosx_10_9_${stdenv.targetPlatform.linuxArch}"
```
Comment (samuela) [inline pkgs/development/python-modules/jaxlib/default.nix:272]: Why `macosx_10_9`? It looks like the pypi wheels are using the `macosx_11_0` tag already: https://pypi.org/project/jaxlib/#files.

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -4501,12 +4501,17 @@ in {
     cudaPackages = pkgs.cudaPackages_11_6;
   };
 
-  jaxlib-build = callPackage ../development/python-modules/jaxlib {
+  jaxlib-build = callPackage ../development/python-modules/jaxlib rec {
+    inherit (pkgs.darwin) cctools;
+    buildBazelPackage = pkgs.buildBazelPackage.override {
+      stdenv = pkgs.darwin.apple_sdk_11_0.stdenv;
+    };
```
Comment (samuela) [inline pkgs/top-level/python-packages.nix:4508]: This seems like the kind of thing we should only do on Darwin?

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 6]
```diff
@@ -199,26 +203,24 @@ let
...
+    bazelFlags = [
       "-c opt"
     ] ++ lib.optional (stdenv.targetPlatform.isx86_64 && stdenv.targetPlatform.isUnix) [
       "--config=avx_posix"
     ] ++ lib.optional cudaSupport [
       "--config=cuda"
     ] ++ lib.optional mklSupport [
       "--config=mkl_open_source_only"
+    ] ++ lib.optionals stdenv.cc.isClang [
+      "--cxxopt=-x" "--cxxopt=c++" "--host_cxxopt=-x" "--host_cxxopt=c++"
```
Comment (samuela) [inline pkgs/development/python-modules/jaxlib/default.nix:215]: OOC why were these flags chosen?

**Review by samuela — COMMENTED** [review]
Thanks for contributing this! Getting jaxlib support on Darwin would be a big win! 

To some extent I'm flying blind here without access to x86_64-darwin hardware, but I have a few questions first anyhow...

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 4 above)
Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:272]: note to self: we should also update to `manylinux_2014` as it seems that's what upstream is doing nowadays as well

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 2 above)
Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:206]: `bazelFlags` are inherited by both the build (https://github.com/NixOS/nixpkgs/blob/7b5ac19d7b5f6e029001b101cab7b73b7a45ed71/pkgs/build-support/build-bazel-package/default.nix#L55) and the fetch (https://github.com/NixOS/nixpkgs/blob/7b5ac19d7b5f6e029001b101cab7b73b7a45ed71/pkgs/build-support/build-bazel-package/default.nix#L59) derivations.

Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:206]: Gotcha, thanks for explaining. In that case, could we keep that comment -- or something analogous -- explaining why we need those flags passed to both steps?

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 4 above)
Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:272]: That comes from https://github.com/google/jax/blob/jaxlib-v0.3.0/build/build_wheel.py#L261-L262, looks like `macosx_11_0` is only used for arm on `0.3.0`.

File: pkgs/top-level/python-packages.nix (hunk 5 above)
Comment (uri-canva) [reply pkgs/top-level/python-packages.nix:4508]: Yeah you're right, https://nixos.org/manual/nixpkgs/unstable/#sec-darwin documents that this is fine:
> On Linux, this will have the same effect as pkgs.callPackage, so you can use pkgs.darwin.apple_sdk_11_0.callPackage regardless of platform.

but it's unintuitive enough that I'll make this conditional.

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 4 above)
Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:272]: ah I see... very good catch. And I guess for the time being we will only support x86_64-darwin anyhow?

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 6 above)
Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:215]: This is due to https://github.com/NixOS/nixpkgs/issues/150655, it's a workaround that forces all files to be compiled as c++ files, I'll add a comment.

File: pkgs/top-level/python-packages.nix (hunk 5 above)
Comment (samuela) [reply pkgs/top-level/python-packages.nix:4508]: ah I see. yeah I think sticking to a conditional given the unintuitive-ness is a good idea

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 3 above)
Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:258]: Both are using `pkgs.protobuf`, we pass `com_google_protobuf` in `TF_SYSTEM_LIBS` to configure the build to use protobuf libraries and headers from the system (https://github.com/tensorflow/tensorflow/blob/144d3dc0929569224728c7634ccd93c8a1bd3da9/third_party/repo.bzl#L66), by stubbing out the vendored build targets with targets that only set compiler and linker flags (https://github.com/tensorflow/tensorflow/blob/144d3dc0929569224728c7634ccd93c8a1bd3da9/third_party/systemlibs/protobuf.BUILD#L62-L78).

It's true that by hardcoding the path here it makes it impossible for hooks to affect it, compared to what we do for gcc, but hopefully it's ok.

**Comment by uri-canva** [comment]
Addressed review.

**Comment by uri-canva** [comment]
Fixed it on `aarch64-darwin` too, so you can test it locally now.

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 7]
```diff
@@ -232,17 +239,31 @@ let
...
           sed -i 's@include/pybind11@pybind11@g' $src
         done
-        sed -i 's@-lprotobuf@-l:libprotobuf.a@' ../output/external/org_tensorflow/third_party/systemlibs/protobuf.BUILD
-        sed -i 's@-lprotoc@-l:libprotoc.a@' ../output/external/org_tensorflow/third_party/systemlibs/protobuf.BUILD
         substituteInPlace ../output/external/org_tensorflow/tensorflow/compiler/xla/python/pprof_profile_builder.cc \
           --replace "status.message()" "std::string{status.message()}"
+        substituteInPlace ../output/external/rules_cc/cc/private/toolchain/osx_cc_wrapper.sh.tpl \
+          --replace "/usr/bin/install_name_tool" "${cctools}/bin/install_name_tool"
+        substituteInPlace ../output/external/rules_cc/cc/private/toolchain/unix_cc_configure.bzl \
+          --replace "/usr/bin/libtool" "${cctools}/bin/libtool"
```
Comment (uri-canva) [inline pkgs/development/python-modules/jaxlib/default.nix:253]: I put some of this patching that is compatible with all platforms in the common block, so the patching can be easily tested on all platforms when updating.

**Comment by samuela** [comment]
Sadly it seems that darwin builds may be re-broken until we can get macOS SDK 10.14 into nixpkgs (https://github.com/NixOS/nixpkgs/pull/183051) 😞

**Comment by uri-canva** [comment]
That should be ok at least from an SDK point of view, we're using the macOS 11.0 SDK here, that's what the `pkgs.darwin.apple_sdk_11_0.stdenv` is about. Hopefully it will just work or require only small changes.

**Comment by uri-canva** [comment]
Does anyone know why ofborg won't pick up `python3Packages.jaxlib` from the PR title automatically?

**Comment by uri-canva** [comment]
@ofborg build python3Packages.jaxlib python39Packages.jaxlib

**Comment by uri-canva** [comment]
Would also be good to figure out why the linux derivations fail to evaluate on ofborg, they build properly in hydra.

**Comment by uri-canva** [comment]
https://github.com/ofborg build python3Packages.jaxlib

**Comment by uri-canva** [comment]
@ofborg build python3Packages.jaxlib

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 184395` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>34 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.aeppl</li>
    <li>python310Packages.aesara</li>
    <li>python310Packages.arviz</li>
    <li>python310Packages.augmax</li>
    <li>python310Packages.chex</li>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.distrax</li>
    <li>python310Packages.dm-sonnet</li>
    <li>python310Packages.elegy</li>
    <li>python310Packages.flax</li>
    <li>python310Packages.objax</li>
    <li>python310Packages.optax</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.rlax</li>
    <li>python310Packages.tensorflow-datasets</li>
    <li>python310Packages.treex</li>
    <li>python310Packages.vqgan-jax</li>
    <li>python39Packages.aeppl</li>
    <li>python39Packages.aesara</li>
    <li>python39Packages.arviz</li>
    <li>python39Packages.augmax</li>
    <li>python39Packages.chex</li>
    <li>python39Packages.dalle-mini</li>
    <li>python39Packages.distrax</li>
    <li>python39Packages.dm-sonnet</li>
    <li>python39Packages.elegy</li>
    <li>python39Packages.flax</li>
    <li>python39Packages.objax</li>
    <li>python39Packages.optax</li>
    <li>python39Packages.pymc</li>
    <li>python39Packages.rlax</li>
    <li>python39Packages.tensorflow-datasets</li>
    <li>python39Packages.treex</li>
    <li>python39Packages.vqgan-jax</li>
  </ul>
</details>
<details>
  <summary>10 packages failed to build:</summary>
  <ul>
    <li>python310Packages.dm-haiku</li>
    <li>python310Packages.jax</li>
    <li>python310Packages.jmp</li>
    <li>python310Packages.numpyro</li>
    <li>python310Packages.treeo</li>
    <li>python39Packages.dm-haiku</li>
    <li>python39Packages.jax</li>
    <li>python39Packages.jmp</li>
    <li>python39Packages.numpyro</li>
    <li>python39Packages.treeo</li>
  </ul>
</details>

**Comment by samuela** [comment]
> Fixed it on `aarch64-darwin` too, so you can test it locally now.

Hmm I'm having trouble building on aarch64-darwin: https://gist.github.com/samuela/471a67182e4c7dded61675fba72f18cf. Looks like it may be a 404

**Comment by uri-canva** [comment]
That's a red herring, it's strange that it's not available on the bazel mirror, but it will get it from github: https://github.com/bazelbuild/rules_cc/commit/081771d4a0e9d7d3aa0eed2ef389fa4700dfb23e.

The error is:
```
ERROR: /private/tmp/nix-build-bazel-build-jaxlib-0.3.0-deps.tar.gz.drv-0/output/external/local_config_cc/BUILD:48:19: in cc_toolchain_suite rule @local_config_cc//:toolchain: cc_toolchain_suite '@local_config_cc//:toolchain' does not contain a toolchain for cpu 'darwin_arm64'
```

Are you sure you checked out the latest commit? That's what the patch file is supposed to fix.

**Comment by uri-canva** [comment]
I'm seeing the same error on ofborg, will try to repro, there might be something wrong with my own setup.

**Comment by uri-canva** [comment]
But maybe let's merge the working `x86_64-darwin` build in the meantime.

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 8]
```diff
@@ -232,17 +239,31 @@ let
       # 4) Patch tensorflow sources to work with later versions of protobuf. See
       #    https://github.com/google/jax/issues/9534. Note that this should be
       #    removed on the next release after 0.3.0.
+      # 5) Patch in https://github.com/bazelbuild/rules_cc/pull/136 to fix the bazel compiler detection on aarch64-darwin
```
Comment (samuela) [inline pkgs/development/python-modules/jaxlib/default.nix:242]: Thanks for filling this upstream! Since that's done, let's just pull it via `fetchpatch` here instead of copying the patch file into nixpkgs's tree.

Alternatively we can remove it for now since AFAIU the aarch64-darwin build is not yet functional anyhow?

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 3 above)
Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:265]: This is subtle enough that I think it may deserve a comment, but I don't think it's necessarily a blocker for getting this merged.

**Review by samuela — COMMENTED** [review]
Hey @uri-canva, apologies for the delay! I have a lingering comment about using `fetchpatch` and one place where I think a comment might be useful. But after this I think we should be good to merge!

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 8 above)
Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:242]: I don't think `fetchpatch` will work: the commit is on a PR, so when the PR is merged / closed and the branch deleted it will eventually be garbage collected. I'll remove it then.

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 3 above)
Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:263]: Updated the existing comment to clarify this bit.

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 9]
```diff
@@ -55,8 +57,11 @@ let
     homepage = "https://github.com/google/jax";
     license = licenses.asl20;
     maintainers = with maintainers; [ ndl ];
-    platforms = [ "x86_64-linux" "aarch64-darwin" "x86_64-darwin"];
-    hydraPlatforms = ["x86_64-linux" ]; # Don't think anybody is checking the darwin builds
+    platforms = platforms.unix;
+    # aarch64-darwin is broken because of https://github.com/bazelbuild/rules_cc/pull/136
+    # however even with that fix applied, it doesn't work for everyone:
+    # https://github.com/NixOS/nixpkgs/pull/184395#issuecomment-1207287129
+    broken = stdenv.isAarch64;
```
Comment (uri-canva) [inline pkgs/development/python-modules/jaxlib/default.nix:64]: Updated the meta to reflect the new state of platform support. Given that I got it working on `aarch64-darwin` (albeit only on my machine) I assume that not building on `aarch64-linux` is because something is broken, not because it's not supported.

Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:64]: Maybe it even builds correctly, I haven't tried.

**Comment by uri-canva** [comment]
@ofborg build python3Packages.jaxlib

**Comment by uri-canva** [comment]
The failure on ofborg is:
```
ERROR: /private/tmp/nix-build-bazel-build-jaxlib-0.3.0.drv-0/output/external/org_tensorflow/tensorflow/compiler/mlir/hlo/BUILD:400:18: TdGenerate external/org_tensorflow/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/mhlo_canonicalize.inc failed: (Illegal instruction): mlir-tblgen failed: error executing command bazel-out/darwin-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen -gen-rewriters external/org_tensorflow/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/mhlo_canonicalize.td -I ... (remaining 17 arguments skipped)
```

I've had that locally when nix was confused on what the current system architecture was, because it was running on an M1 with rosetta, but I'm not sure why that would happen on an x86_64 mac. I've seen in it some of the past builds too, but not all, so it seems to only affect specific machines.

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 8 above)
Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix]: This is a common pattern in other packages in nixpkgs. It's generally preferred to fetchpatch commits rather than PRs/branches, since PRs are mutable. That's an interesting question re commits being deleted... I'm not sure how that's handled. Perhaps they just link to privately controlled forks of repos? Idk...

**Comment by samuela** [comment]
> I've had that locally when nix was confused on what the current system architecture was, because it was running on an M1 with rosetta, but I'm not sure why that would happen on an x86_64 mac. I've seen in it some of the past builds too, but not all, so it seems to only affect specific machines.

Huh, very odd... doesn't look like your fault in any case. I would be curious to figure out the underlying issue, but doesn't need to be in scope for this particular PR.

**Comment by samuela** [comment]
Thanks for all your hard work on this @uri-canva! These changes LGTM and are a big leap forward in terms of our ability to support more platforms.

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 10]
```diff
@@ -236,13 +246,25 @@ let
...
           sed -i 's@include/pybind11@pybind11@g' $src
         done
-        sed -i 's@-lprotobuf@-l:libprotobuf.a@' ../output/external/org_tensorflow/third_party/systemlibs/protobuf.BUILD
-        sed -i 's@-lprotoc@-l:libprotoc.a@' ../output/external/org_tensorflow/third_party/systemlibs/protobuf.BUILD
         substituteInPlace ../output/external/org_tensorflow/tensorflow/compiler/xla/python/pprof_profile_builder.cc \
           --replace "status.message()" "std::string{status.message()}"
+        substituteInPlace ../output/external/rules_cc/cc/private/toolchain/osx_cc_wrapper.sh.tpl \
+          --replace "/usr/bin/install_name_tool" "${cctools}/bin/install_name_tool"
+        substituteInPlace ../output/external/rules_cc/cc/private/toolchain/unix_cc_configure.bzl \
+          --replace "/usr/bin/libtool" "${cctools}/bin/libtool"
```
Comment (andersk) [inline pkgs/development/python-modules/jaxlib/default.nix:254]: This broke `jaxlib` on non-Darwin because `cctools` is Darwin-specific.

https://github.com/NixOS/nixpkgs/blob/4dc0fce489d180acb37a52f6569068175b9a7954/pkgs/os-specific/darwin/cctools/port.nix#L100

Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:254]: Unfortunate, it wasn't evaluating on ofborg so I couldn't check, have you already fixed it? I'll put up a fix.

Comment (uri-canva) [reply pkgs/development/python-modules/jaxlib/default.nix:254]: https://github.com/NixOS/nixpkgs/pull/187536
